noir/noir-repo/aztec_macros/src/transforms/note_interface.rs

```diff
@@ -0,0 +1,510 @@
+    noir_fn.def.visibility = ItemVisibility::Public;
+    Ok(noir_fn)
+}
+
+fn generate_note_serialize_content(
+    note_type: &str,
+    note_fields: &[(String, String)],
+    note_serialize_len: &String,
+    impl_span: Option<Span>,
+) -> Result<NoirFunction, AztecMacroError> {
+    let function_source =
+        generate_note_serialize_content_source(note_type, note_fields, note_serialize_len);
+
+    let (function_ast, errors) = parse_program(&function_source);
+    if !errors.is_empty() {
+        return Err(AztecMacroError::CouldNotImplementNoteSerialization {
+            secondary_message: Some("Failed to parse Noir macro code (fn seserialize_content). This is either a bug in the compiler or the Noir macro code".to_string()),
+            span: impl_span
+        });
+    }
+
+    let mut function_ast = function_ast.into_sorted();
+    let mut noir_fn = function_ast.functions.remove(0);
+    noir_fn.def.visibility = ItemVisibility::Public;
+    Ok(noir_fn)
+}
+
+fn generate_note_fields_fn(
+    note_type: &str,
+    note_fields: &[(String, String)],
+    impl_span: Option<Span>,
+) -> Result<NoirFunction, AztecMacroError> {
+    let function_source = generate_note_fields_fn_source(note_type, note_fields);
+    let (function_ast, errors) = parse_program(&function_source);
+    if !errors.is_empty() {
+        return Err(AztecMacroError::CouldNotImplementNoteSerialization {
+            secondary_message: Some("Failed to parse Noir macro code (fn fields). This is either a bug in the compiler or the Noir macro code".to_string()),
+            span: impl_span
+        });
+    }
+    let mut function_ast = function_ast.into_sorted();
+    let mut noir_fn = function_ast.functions.remove(0);
+    noir_fn.def.visibility = ItemVisibility::Public;
+    Ok(noir_fn)
+}
+
+fn generate_compute_note_content_hash(
+    note_type: &String,
+    impl_span: Option<Span>,
+) -> Result<NoirFunction, AztecMacroError> {
+    let function_source = format!(
+        "
+        fn compute_note_content_hash(self: {}) -> Field {{
+            // TODO(#1205) Should use a non-zero generator index.
+            dep::aztec::hash::pedersen_hash(self.serialize_content(), 0)
+        }}
+        ",
+        note_type
+    );
+    let (function_ast, errors) = parse_program(&function_source);
+    if !errors.is_empty() {
+        return Err(AztecMacroError::CouldNotImplementNoteSerialization {
+            secondary_message: Some("Failed to parse Noir macro code (fn compute_note_content_hash). This is either a bug in the compiler or the Noir macro code".to_string()),
+            span: impl_span
+        });
+    }
+    let mut function_ast = function_ast.into_sorted();
+    let mut noir_fn = function_ast.functions.remove(0);
+    noir_fn.def.visibility = ItemVisibility::Public;
+    Ok(noir_fn)
+}
+
+fn generate_note_fields_struct_source(note_type: &str, note_fields: &[(String, String)]) -> String {
+    let note_field_selectors = note_fields
```

Comment:
I''d include a comment showing what you're trying to achieve here. Otherwise the reader needs to work backwards from the impl to figure it out.

(and the reader may not even be proficient in Rust)

Reply:
Actually not just this but all of the generate source functions

Reply:
I added comments just before each call in the `generate_note_interface_impl` method because I thought it would be easier to follow. I can move them to their impls instead.

Reply:
Improved docs and comments!